File: java/robot-name/src/main/java/Robot.java
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class Robot {
    private static final Random RANDOM = new Random();
    private static final Set<String> NAMES = new HashSet<>();

    private String name;

    public String getName() {
        if (name == null) {
            name = generateNewName();
        }
        return name;
    }

    public void reset() {
        name = null;
    }

    private static String generateNewName() {
        String name = null;
        while (name == null) {
            name = String.format("%c%c%d%d%d", randomChar(), randomChar(), randomDigit(), randomDigit(), randomDigit());
            if (NAMES.contains(name)) {
                name = null;
            }
        }
        NAMES.add(name);
        return name;
    }

    private static char randomChar() {
        return (char) (((int) 'A') + RANDOM.nextInt(26));
    }

    private static int randomDigit() {
        return RANDOM.nextInt(10);
    }
}

File: kotlin/robot-name/src/main/kotlin/Robot.kt
import kotlin.random.Random

class Robot {
    companion object {
        private val names = mutableSetOf<String>()
        
        private fun newName(): String {
            var name = randomName()
            while (names.contains(name)) {
                name = randomName()
            }
            names.add(name)
            return name
        }

        private fun randomName(): String {
            val sb = StringBuilder()
            sb.append('A' + Random.nextInt(26))
            sb.append('A' + Random.nextInt(26))
            sb.append('0' + Random.nextInt(10))
            sb.append('0' + Random.nextInt(10))
            sb.append('0' + Random.nextInt(10))
            return sb.toString()
        }
    }

    var name: String = newName()
        private set
    
    fun reset() {
        name = newName()
    }
}
